### PR TITLE
Task 2

### PR DESCRIPTION
**Pull request by Pl-Alex**

### CODE REVIEW

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -24,4 +26,6 @@ junit = { group = "junit", name = "junit", version.ref = "junit" }
 testExtJUnit = { group = "androidx.test.ext", name = "junit", version.ref = "testExtJUnit" }
 espressoCore = { group = "androidx.test.espresso", name = "espresso-core", version.ref = "espresso" }
 datastorePreferences = { group = "androidx.datastore", name = "datastore-preferences", version.ref = "datastore" }
+cardView = { group = "androidx.cardview", name = "cardview", version.ref = "cardview" }
+javaFaker = { group = "com.github.javafaker", name = "javafaker", version.ref = "javafaker"}
 [plugins]
```
Comment (1937Carrie) [inline gradle/libs.versions.toml]: There is empty the plugins part. Look [this](https://developer.android.com/build/migrate-to-catalogs#migrate-plugins) or [this](https://youtu.be/MWw1jcwPK3Q?feature=shared).

File: app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+import com.alexP.assignment1.viewModels.ContactsViewModel
+
+class ContactsLoader(private val context: Context, private val viewModel: ContactsViewModel) {
+
+    companion object {
+        private const val PERMISSIONS_REQUEST_READ_CONTACTS = 100
+    }
+
+    fun loadContacts() {
+        if (ContextCompat.checkSelfPermission(
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt]: Using onRequestPermissionsResult() in AppCompatActivity is indeed possible ([proof](https://developer.android.com/reference/androidx/core/app/ActivityCompat.OnRequestPermissionsResultCallback#onRequestPermissionsResult(int,java.lang.String[],int[]))). But prepare to level 3 and read [this](https://developer.android.com/reference/androidx/fragment/app/Fragment#onRequestPermissionsResult(int,java.lang.String[],int[])) and [this](https://developer.android.com/training/permissions/requesting#request-permission).

Also, it is better to check the permission in activity or fragment, but based on result of requesting permissions, call the appropriate function from viewmodel or do some another stuff.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+import com.alexP.assignment1.R
+import com.alexP.assignment1.databinding.ItemContactBinding
+import com.alexP.assignment1.model.Contact
+import com.bumptech.glide.Glide
+import com.bumptech.glide.request.RequestOptions
+
+interface ContactActionListener {
+
+    fun onContactDelete(contact: Contact)
+}
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: Better extract to separate class.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+    override fun areContentsTheSame(oldItem: Contact, newItem: Contact): Boolean {
+        return oldItem == newItem
+    }
+
+}
+
+class ContactsAdapter(
+    private val userActionListener: ContactActionListener
+) : ListAdapter<Contact, ContactsAdapter.ContactsViewHolder>(ContactsDiffCallback()), View.OnClickListener {
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: You are not going to need to inherit from View.OnClickListener.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ContactsViewHolder {
+        val inflater = LayoutInflater.from(parent.context)
+        val binding = ItemContactBinding.inflate(inflater, parent, false)
+
+        binding.root.setOnClickListener(this)
+        binding.buttonTrash.setOnClickListener(this)
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: Get rid of these.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            buttonTrash.tag = contact
+
+            textViewContactFullName.text = contact.fullName
+            textViewContactCareer.text = contact.career
+            if (contact.photo.isNotBlank()) {
+                imageViewContactImage.loadCircularImage(contact.photo)
+            } else {
+                imageViewContactImage.setImageResource(R.drawable.default_contact_image)
+            }
+        }
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: In onBindViewHolder() function just call holder' bind() function.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+    override fun onClick(v: View) {
+        val contact = v.tag as Contact
+        when (v.id) {
+            R.id.button_trash -> {
+                userActionListener.onContactDelete(contact)
+            }
+        }
+
+    }
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: Get rid of this.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                imageViewContactImage.loadCircularImage(contact.photo)
+            } else {
+                imageViewContactImage.setImageResource(R.drawable.default_contact_image)
+            }
+        }
+    }
+
+    class ContactsViewHolder(
+        val binding: ItemContactBinding
+    ) : RecyclerView.ViewHolder(binding.root)
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: Create bind() function and use it scope for configuring holder view. I.e. move old code from onBindViewHolder() to this function and also set on click listeners in this function. General rule for setting listeners in adapter: you should declare all instructions in activity or fragment class and pass them to adapter but use id of item from recycler adapter.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+    }
+}
+
+fun ImageView.loadCircularImage(imageLink: String) {
+    Glide.with(this).load(imageLink).apply(RequestOptions.circleCropTransform())
+        .placeholder(R.drawable.default_contact_image).error(R.drawable.default_contact_image)
+        .into(this)
+}
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: This is look like function for common util purposes. Extract her from adapter class to "util" class.

File: app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+import android.content.Context
+import android.content.pm.PackageManager
+import android.provider.ContactsContract
+import androidx.appcompat.app.AppCompatActivity
+import androidx.core.app.ActivityCompat
+import androidx.core.content.ContextCompat
+import com.alexP.assignment1.model.Contact
+import com.alexP.assignment1.viewModels.ContactsViewModel
+
+class ContactsLoader(private val context: Context, private val viewModel: ContactsViewModel) {
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt]: I think it is bad idea to pass viewmodel instance anywhere. Consult with mates in discord about using usecases here (because i do not know how to use it) or you can do all this stuff in viewmodel directly.

File: app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+                arrayOf(Manifest.permission.READ_CONTACTS),
+                PERMISSIONS_REQUEST_READ_CONTACTS
+            )
+        } else {
+            fetchContacts()
+        }
+    }
+
+    private fun fetchContacts() {
+        val resolver: ContentResolver = context.contentResolver
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/dataProviders/ContactsLoader.kt]: It is better to pass instance of ContentResolver directly to fetchContracts() function parameters than pass it to class properties.

File: app/src/main/java/com/alexP/assignment1/model/ContactsService.kt [hunk 12]
```diff
@@ -0,0 +1,77 @@
+package com.alexP.assignment1.model
+
+
+import com.github.javafaker.Faker
+
+typealias ContactsListener = (contacts: List<Contact>) -> Unit
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/model/ContactsService.kt]: Don't understand about the use of this typealias. Please explain at free time.

File: app/src/main/java/com/alexP/assignment1/ui/AddContactFragment.kt [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+import com.alexP.assignment1.utils.validator.base.BaseValidator
+import com.bumptech.glide.Glide
+import com.bumptech.glide.request.RequestOptions
+
+
+class AddContactFragment : DialogFragment() {
+
+    interface OnContactSavedListener {
+        fun onContactSaved(contact: Contact)
+    }
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/AddContactFragment.kt]: Better extract to separate class.

File: app/src/main/java/com/alexP/assignment1/ui/BaseActivity.kt
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/BaseActivity.kt:1]: It is better to extract this class to separate package

File: app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+import com.alexP.assignment1.dataProviders.ContactsLoader
+import com.alexP.assignment1.databinding.ActivityContactsBinding
+import com.alexP.assignment1.model.Contact
+import com.alexP.assignment1.ui.utils.SpacingItemDecorator
+import com.alexP.assignment1.ui.utils.factory
+import com.alexP.assignment1.viewModels.ContactsViewModel
+import com.google.android.material.snackbar.Snackbar
+
+
+class ContactsActivity : AppCompatActivity(), AddContactFragment.OnContactSavedListener {
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt]: It's better to create a property with an interface type rather than inherit from it.

File: app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+        val itemAnimator = binding.recyclerView.itemAnimator
+        if (itemAnimator is DefaultItemAnimator) {
+            itemAnimator.supportsChangeAnimations = false
+        }
+    }
+
+    private fun showAddContactDialog() {
+        val fragmentManager = supportFragmentManager
+        val fragment = AddContactFragment()
+        fragment.setOnContactSavedListener(this)
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt]: And then you can directly pass the implementation of the interface, and not the entire activity (or fragment).

File: app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+            fragment.show(fragmentManager, "dialog")
+        } else {
+            val transaction = fragmentManager.beginTransaction()
+            transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
+            transaction
+                .add(android.R.id.content, fragment)
+                .addToBackStack(null)
+                .commit()
+        }
+    }
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/ContactsActivity.kt]: Please explain me how it works (in discord voicechat at free time), i never seen this before.

File: app/src/main/java/com/alexP/assignment1/ui/utils/ViewModelFactory.kt [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+
+            else -> {
+                throw IllegalStateException("Unknown view model class")
+            }
+        }
+        return viewModel as T
+    }
+}
+
+fun Activity.factory() = ViewModelFactory(applicationContext as App)
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/utils/ViewModelFactory.kt]: Ideally, you should have a viewmodel for each activity or fragment to be ready execute business logic in them, but this ext function will be able create factory only for ContactsViewModel.
Also check out this [information](https://developer.android.com/topic/libraries/architecture/viewmodel/viewmodel-cheatsheet).

**Review by 1937Carrie — CHANGES_REQUESTED** [review]

File: contactsprovider/data/src/main/java/com/alexp/contactsprovider/data/Contact.kt [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+
+data class Contact(
+    val id: Long,
+    val photo: String,
+    val fullName: String,
+    val career: String,
+    val email: String,
+    val phone: String,
+    val address: String,
+    val dateOfBirth: String
```
Comment (1937Carrie) [inline contactsprovider/data/src/main/java/com/alexp/contactsprovider/data/Contact.kt]: Check out [material](https://kotlinlang.org/docs/coding-conventions.html#trailing-commas) about trailing commas and reformat all your code.

File: contactsprovider/data/src/main/AndroidManifest.xml
Comment (1937Carrie) [inline contactsprovider/data/src/main/AndroidManifest.xml:1]: Collapse empty tag. At the review, can tell your mentor that you optimized the file size by reducing it by 17,74% (62 B -> 51 B). Make it for all new manifests.

File: datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+        }
+    }
+
+    suspend fun saveCredentials(email: String, password: String) {
+        saveRememberMeState()
+        saveString(EMAIL, email)
+        saveString(PASSWORD, password)
+    }
+
+    suspend fun saveBoolean(key: String, value: Boolean) {
```
Comment (1937Carrie) [inline datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt]: Make all highlighted functions private.

File: app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt [hunk 20]
```diff
@@ -0,0 +1,149 @@
...
+import com.alexp.textvalidation.data.validator.base.ValidationResult
+import kotlinx.coroutines.launch
+
+
+class AddContactFragment(
+    private val onSaveAction: (Contact) -> Unit
+) : DialogFragment() {
+
+    private lateinit var binding: FragmentDialogAddContactBinding
+    private lateinit var viewModel: AddContactViewModel
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt]: Now that you've practiced creating a model instance manually, let's take a look at some other [ViewModel APIs](https://developer.android.com/topic/libraries/architecture/viewmodel/viewmodel-cheatsheet) that make creating a model easier.

File: app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt [hunk 21]
```diff
@@ -0,0 +1,149 @@
...
+            }
+        }
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View {
+        binding = FragmentDialogAddContactBinding.inflate(inflater, container, false)
+        binding.root.applyWindowInsets()
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt:48]: It is good idea using extension function for displaying content edge-to-edge but a little bit wrong implementation. Remind me at you free time, I will show you what is wrong and how to handle it.

File: app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt [hunk 22]
```diff
@@ -0,0 +1,149 @@
...
+        toolbar.setNavigationOnClickListener {
+            dismiss()
+        }
+        setListeners()
+
+        return binding.root
+    }
+
+    private fun observeValues() {
+        lifecycleScope.launch {
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt]: Using this approach in fragments may cause to app crashing. Add following code, run program, come in and come out by using nav icon on toolbar from that fragment. Think about why the crash occurred.
```kotlin
override fun onDestroyView() {
        super.onDestroyView()
        i = null
    }

private var i: Int? = 0
private fun observeValues() {
    lifecycleScope.launch {
        while (true) {
            delay(250L)
            i = i!! + 1
            binding.topBar.title = i.toString()
        }
        viewModel.galleryUri.collect { uri ->
            uri?.let {
                binding.imageViewProfileImage.loadCircularImage(it.toString())
            }
        }
    }
}
```

File: app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+        savedInstanceState: Bundle?
+    ): View {
+        binding = FragmentDialogAddContactBinding.inflate(inflater, container, false)
+        binding.root.applyWindowInsets()
+
+        viewModel = ViewModelProvider(this)[AddContactViewModel::class.java]
+
+        observeValues()
+        val toolbar = binding.topBar
+        toolbar.setNavigationOnClickListener {
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt]: Concatenate val toolbar declaration and setNavigationOnClickListener function call.

File: app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+            email = binding.inputEditTextEmail.text.toString(),
+            phone = binding.inputEditTextPhone.text.toString(),
+            address = binding.inputEditTextAddress.text.toString(),
+            dateOfBirth = binding.inputEditTextDateOfBirth.text.toString()
+        )
+        onSaveAction(contact)
+        dismiss()
+    }
+
+    private fun enteredDataIsInvalid(): Boolean {
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt]: This function is performing asserting inputted string is valid. Things that are not related to UI layer have to be extracted to other layers and be calling from viewmodel.

File: app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt [hunk 25]
```diff
@@ -29,49 +29,34 @@ class MyProfileActivity : BaseActivity<ActivityMyProfileBinding>() {
...
-        lifecycleScope.launch {
-            val email = readEmail(dataStore)
-            binding?.textViewNameSurname?.text = parseEmail(email!!)
-        }
+        viewModel =
+            ViewModelProvider(this, MyProfileViewModelFactory(this))[MyProfileViewModel::class.java]
 
-        binding?.let {
+        binding.let {
             Glide.with(this)
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt]: You already have function ImageView.loadCircularImage(). Replace Glide snippet with your extension function.

**Review by 1937Carrie — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+
+    override fun areItemsTheSame(oldItem: Contact, newItem: Contact): Boolean {
+        return oldItem.id == newItem.id
+    }
+
+    override fun areContentsTheSame(oldItem: Contact, newItem: Contact): Boolean {
+        return oldItem == newItem
+    }
+
+}
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: I think it is better to move ContactsDiffCallback class implementation below ContactsAdapter class, because this Kotlin file is called ContactsAdapter, hense main object is ContactsAdapter. Related to entire project.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 27]
```diff
@@ -0,0 +1,77 @@
...
+        holder.bind(getItem(position))
+    }
+
+    class ContactsViewHolder(
+        private val binding: ItemContactBinding,
+        private val userActionListener: IContactActionListener,
+    ) : RecyclerView.ViewHolder(binding.root) {
+        fun bind(contact: Contact) {
+            with(binding) {
+                itemView.tag = contact
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: For lines 48, 49, 68 and related to them, I think it is better to avoid using tags and implement clickListeners directly in setOnClickListener() function for each element respectively.

File: app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthActivity.kt
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthActivity.kt:1]: Handling for password textInputLayout error description is missing.

File: app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthViewModel.kt [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+        return namePart.split('.')
+            .joinToString(" ") { it ->
+                it.replaceFirstChar {
+                    if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString()
+                }
+            }
+    }
+
+    companion object {
+        fun createFactory(context: Context): ViewModelProvider.Factory = viewModelFactory {
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthViewModel.kt]: I made a mistake using this approach. You should create a DataStoreProvider instance outside the factory and pass that instance to factory. So the createFactory() function should take a DataStoreProvider instance, not a Context. Related to entire project.

File: app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt [hunk 29]
```diff
@@ -0,0 +1,182 @@
...
+import com.alexP.assignment1.ui.contactsActivity.ContactsViewModel
+import com.alexP.assignment1.ui.utils.SpacingItemDecorator
+import com.alexp.contactsprovider.data.Contact
+import com.google.android.material.snackbar.Snackbar
+
+
+class ContactsActivity : BaseActivity<ActivityContactsBinding>() {
+    private lateinit var adapter: ContactsAdapter
+
+    private lateinit var viewModel: ContactsViewModel
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt]: Its time to get rid of lateinit var viewModel declaration and change it using viewModels() api, e.g. like you have already done in AddContactFragment. Related to entire project.

File: app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+                Snackbar.make(
+                    this,
+                    binding.root,
+                    getString(R.string.permission_not_granted),
+                    Snackbar.LENGTH_SHORT
+                ).show()
+            }
+        }
+
+    private val onSaveAction: (Contact) -> Unit =
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt]: It is better change to style like
```kotlin
private val onSaveAction: (Contact) -> Unit = { contact: Contact ->
    viewModel.addContact(contact)
}
```

File: app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt [hunk 31]
```diff
@@ -0,0 +1,61 @@
...
+import com.alexP.assignment1.ui.authActivity.AuthActivity
+import com.alexP.assignment1.utils.loadCircularImage
+import com.alexP.assignment1.utils.parseEmail
+import kotlinx.coroutines.launch
+
+class MyProfileActivity : BaseActivity<ActivityMyProfileBinding>() {
+
+    private lateinit var viewModel: MyProfileViewModel
+
+    companion object {
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt]: Move companion object to the bottom of class. [Source](https://kotlinlang.org/docs/coding-conventions.html#class-layout).

File: app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileActivity.kt:1]: Handling for intent parameter(email) is missing.

**Review by 1937Carrie — CHANGES_REQUESTED** [review]
As a result of the architecture redesign, the display of errors in the password textInputLayout, and the receipt and processing of intent parameter have missed - need to bring back it. There are some style mistakes; mistake when working with recyclerview. Overall, it's fine.

File: app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactViewModel.kt [hunk 32]
```diff
@@ -0,0 +1,66 @@
...
+import android.net.Uri
+import androidx.lifecycle.ViewModel
+import com.alexP.assignment1.utils.getValidationResultMessage
+import com.alexp.textvalidation.data.validator.EmailValidator
+import com.alexp.textvalidation.data.validator.EmptyValidator
+import com.alexp.textvalidation.data.validator.base.BaseValidator
+import com.alexp.textvalidation.data.validator.base.ValidationResult
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.asStateFlow
+
```
Comment (Pl-Alex) [inline app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactViewModel.kt:12]: q

File: app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt [hunk 33]
```diff
@@ -0,0 +1,153 @@
...
+            inputEditTextDateOfBirth.setOnFocusChangeListener { _, hasFocus -> if (!hasFocus) validateDateOfBirth() }
+        }
+    }
+
+    private fun validateUsername(): Boolean {
+        val validationResult = vm.validateUsername(binding.inputEditTextUsername.text.toString())
+        binding.inputLayoutUsername.error = validationResult?.let { getString(it) } ?: ""
+        return validationResult == null
+    }
+
```
Comment (Pl-Alex) [inline app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactFragment.kt:86]: q

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 34]
```diff
@@ -43,35 +30,40 @@ class ContactsAdapter(
         private val binding: ItemContactBinding,
         private val userActionListener: IContactActionListener,
     ) : RecyclerView.ViewHolder(binding.root) {
+
+        private lateinit var contact: Contact
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: Get rid of this.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt [hunk 35]
```diff
@@ -43,35 +30,40 @@ class ContactsAdapter(
...
-                R.id.button_trash -> {
+                buttonTrash.setOnClickListener {
                     userActionListener.onContactDelete(contact)
                 }
             }
+        }
 
+        fun getContact(): Contact {
+            return contact
         }
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt]: Get rid of this also.

File: app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/adapters/ContactsAdapter.kt:1]: Move whole content of package adapters to package contactsActivity, because this stuff is used only in the ContactsActivity.
image

File: app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt [hunk 36]
```diff
@@ -113,7 +110,7 @@ class ContactsActivity : BaseActivity<ActivityContactsBinding>() {
             }
 
             override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
-                deleteContact(viewHolder.itemView.tag as Contact)
+                deleteContact((viewHolder as ContactsAdapter.ContactsViewHolder).getContact())
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt]: Since I already said to remove the getContact() function from the ViewHolder, there is another way to accomplish deleting a contact from the list. In this particular case, the order of contacts in the viewmodel field is equal to the order in which contacts are displayed in the recyclerview.
Your code will be look like
```kotlin
val contact = /*retrieve contact using the default stack of fields or properties of passed viewholder*/
deleteContact(contact) // or little bit transformed form that respects contacts nullability
```

File: app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactViewModel.kt
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/addContactFragment/AddContactViewModel.kt:1]: It is bad you removed these functions from viewmodel. As main purpose of these functions is provide some business logic to UI, so these functions must be called from viewmodel since viewmodel is stateholder for UI. So yes, you have to called these function in viewmodel and then called the required viewmodel function in UI class. You're throwing data around between the UI, the viewmodel, and where that logic is defined, but this is the way.

**Review by 1937Carrie — CHANGES_REQUESTED** [review]

File: datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt
Comment (1937Carrie) [inline datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt:1]: You misunderstood. Sign Up screen is imitation of authorization process before fifth level. But for authorization you need both an email and a password. So you either need store email and password together  if checkbox "Remember Me" was checked or don't store anything otherwise. A function saveRememberMeState() is redundant in this case. As a conclusion, when you start your app every next time, you check if datastore have credentials and direct the flow of code execution in the right direction, based on that result.
The behaviour have to be as in any of modern apps such telegram, instagram, gmail, etc. If datastore have credentials, you even do not see authorization screen, in ideally written project and you will immediately be taken to the main screen of the application.

Comment (1937Carrie) [reply datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt:1]: And I think you should rework your DataStore read value functions so they will return Flow\<T\> but not T, because it is implicitly recommended in the [documentation](https://developer.android.com/topic/libraries/architecture/datastore#preferences-read). Now, if i understood correctly, you are writing value to datastore and waiting while value will be successfully written. But you need just write value to data store and wait only when you are reading the value.

File: app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt [hunk 37]
```diff
@@ -110,7 +108,9 @@ class ContactsActivity : BaseActivity<ActivityContactsBinding>() {
             }
 
             override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
-                deleteContact((viewHolder as ContactsAdapter.ContactsViewHolder).getContact())
+                val contact = vm.contacts.value
+                    ?.get((viewHolder as ContactsAdapter.ContactsViewHolder).adapterPosition)
+                contact?.let{ deleteContact(contact)}
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/contactsActivity/ContactsActivity.kt]: The following code will be enough:
```kotlin
val contact = vm.contacts.value?.get(viewHolder.adapterPosition)
contact?.let{ deleteContact(contact)}
```

**Review by 1937Carrie — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthViewModel.kt [hunk 38]
```diff
@@ -23,25 +24,30 @@ class AuthViewModel(
...
 
-    fun saveState(email: String, isRememberMeChecked: Boolean) {
+    fun saveUsername(username: String) {
         viewModelScope.launch {
-            dataStore.saveCredentials(email)
-            dataStore.saveRememberMeState(isRememberMeChecked)
-            updateAuthState()
+            dataStore.saveUsername(username)
         }
     }
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthViewModel.kt:52]: Redundant code

File: app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthActivity.kt [hunk 39]
```diff
@@ -64,9 +65,15 @@ class AuthActivity : BaseActivity<ActivityAuthBinding>() {
     private fun onRegisterButtonPressed() {
         if (isAnyEnteredDataInvalid()) return
         val emailText = binding.inputEditTextEmail?.text.toString().lowercase()
-        val isRememberMeChecked = binding.checkBoxRemember?.isChecked ?: false
-        vm.saveState(emailText, isRememberMeChecked)
-        onNavigate(isRememberMeChecked)
+        val passwordText = binding.inputEditTextPassword?.text.toString()
+        val username = parseEmail(emailText)
+
+        vm.saveUsername(username)
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/authActivity/AuthActivity.kt:71]: Redundant code.

File: app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileViewModel.kt [hunk 40]
```diff
@@ -15,13 +15,15 @@ class MyProfileViewModel(
...
+    val myProfileState get() = _myProfileState.asStateFlow()
 
     init {
         viewModelScope.launch {
-            _emailState.update {
-                dataStore.readEmail()
+            dataStore.readUsername().collect { username ->
+                _myProfileState.update {
+                    MyProfileState(username = username)
+                }
```
Comment (1937Carrie) [inline app/src/main/java/com/alexP/assignment1/ui/myProfileActivity/MyProfileViewModel.kt:26]: Replace readUsername() function with readEmail() and format that string like
```kotlin
val formattedName = email.substringBefore("@").split(".").joinToString(" ")
```

File: datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt [hunk 41]
```diff
@@ -3,69 +3,65 @@ package com.alexp.datastore.data
...
-import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.map
 
 
-private const val EMAIL = "email"
-private const val REMEMBER_STATE = "remember_state"
+private val emailKey = stringPreferencesKey("email")
+private val usernameKey = stringPreferencesKey("username")
```
Comment (1937Carrie) [inline datastore/data/src/main/java/com/alexp/datastore/data/DataStoreProvider.kt:15]: Get rid of this.

**Review by 1937Carrie — APPROVED** [review]
